### DISCUSSION TITLE
Diffent behaviour with variable propagation in BIOS/UEFI case

### DISCUSSION THREAD

**Discussion by v01t** [ipxe/ipxe#1241] [General]
Hello,

For first I'd like to say thank you for a great product, I'm in love :clap: 

I'm having setup where dhcp return proper boot file option (undionly.kpxe, ipxe.efi) depending on firmware, that search for `autoexec.ipxe` in both cases so I define it as follows:
```
#!ipxe
dhcp && goto netboot || goto dhcperror

:dhcperror
prompt --key s --timeout 10000 DHCP failed, hit 's' for the iPXE shell; reboot in 10 seconds && shell || reboot

:netboot
set dns:ipv4 ${net0.dhcp/dns}
set boot-host http://boot.domain
chain --autofree ${boot-host}/ipxe/${cpuvendor}.bin ||
chain ${boot-host}/ipxe/menu.ipxe ||
prompt --key s --timeout 10000 Chainloading failed, hit 's' for the iPXE shell; reboot in 10 seconds && shell || reboot
```
that gets the IP from DHCP, chain vendor specific binaries and then chain `menu.ipxe`, important moment here is `set boot-host http://boot.domain` (variable are set), in `menu.ipxe` no specific things happen besides actual boot, here is main boot part:
```
...
set boot-url ${boot-host}/os/${os}/${os-release}/boot/${arch}
kernel ${boot-url}/vmlinuz-linux
initrd ${boot-url}/initramfs-linux.img
imgargs vmlinuz-linux initrd=initramfs-linux.img boot=live union=overlay archiso_http_srv=${boot-host} archisobasedir=/os/${os}/${os-release} ${extra-imgargs}
boot || goto error
...
```
important moment here is `initargs ... archiso_http_srv=${boot-host}`
with that configuration UEFI firmware based setup boot without issues (including download of airootfs.sfs that located in `${boot-host}/os/${os}/${os-release}/${arch}/airootfs.sfs`), but BIOS based one fail during  pulling of airootfs.sfs via curl

during pull of kernel,initramfs,airootfs in UEFI case all path's starts with `http://boot.domain...` (as expected), but in BIOS case those paths seen as `/os/...` and now most interested and quite misleading part - while such "path" that seen without domain in BIOS only case works during pull of kernel and initramfs, it leads curl to fail during airootfs pull (because `archiso_http_srv=${boot-host}` actually equal to `archiso_http_srv=` in BIOS case)

My solution to make same configuration setup (`autoexec.ipxe` and `menu.ipxe`) work for both UEFI and BIOS was to redefine (again) `boot-host` as `set boot-host http://boot.domain` but in `menu.ipxe` with that also curl was able to pull airootfs in BIOS case.

So while I resolve the issue and most likely do not understand/know/miss something in this case, I still would be interested to hear why this actually happen, as its clear that `boot-host` value actually propagated in both BIOS and UEFI cases to `menu.ipxe` (because kernel and initramfs pulled successfully) but become somehow "unassigned" as adding echo of that variable will show that its empty in BIOS case, but this will not stop boot process to actually pull kernel and initramfs

Thanks.

**Comment by mcb30** [comment]
> For first I'd like to say thank you for a great product, I'm in love 👏

Thank you!

> So while I resolve the issue and most likely do not understand/know/miss something in this case, I still would be interested to hear why this actually happen, as its clear that `boot-host` value actually propagated in both BIOS and UEFI cases to `menu.ipxe` (because kernel and initramfs pulled successfully) but become somehow "unassigned" as adding echo of that variable will show that its empty in BIOS case, but this will not stop boot process to actually pull kernel and initramfs

I have no idea why you are experiencing this.  The variable expansion code is exactly the same in the BIOS and UEFI builds, and there is no reason why you should see any difference in behaviour.

I would start by putting an `imgstat` before your `boot` command, which will give you an opportunity to see what literal command-line has been constructed for your kernel to use.  You can also try using `echo` and `show` to see what values you have in variables at various points in your script.  There is no explanation for the behaviour you are seeing, so this poking around is just to explore what might be happening in more detail.

**Comment by v01t** [comment]
Thanks for your reply @mcb30 

thats exactly what I did to trace the problem, having this boot-host variable set in autoexec.ipxe echo'ing it in menu.ipxe shows thats its empty only in BIOS case, but it affects only underlying OS actions when such variable needs to be propagated futher, like with archiso_http_srv argument that later on used by curl to get airootfs image file.

I can provide exact files if it make sense for future investigation

Regards and thanks again for a great product
